enzyme/Enzyme/MLIR/Interfaces/GradientUtilsReverse.cpp

```diff
@@ -0,0 +1,405 @@
+  Type gradientType = getGradientType(v);
+  OpBuilder initBuilder(initializationBlock, initializationBlock->begin());
+  Value gradient = initBuilder.create<enzyme::CreateCacheOp>(v.getLoc(), gradientType);
+  return gradient;
+}
+
+Value mlir::enzyme::MGradientUtilsReverse::getNewFromOriginal(const mlir::Value originst) const {
+  if (!originalToNewFn.contains(originst)) {
+    llvm::errs() << oldFunc << "\n";
+    llvm::errs() << newFunc << "\n";
+    llvm::errs() << originst << "\n";
+    llvm_unreachable("Could not get new val from original");
+  }
+  return originalToNewFn.lookupOrNull(originst);
+}
+
+Block * mlir::enzyme::MGradientUtilsReverse::getNewFromOriginal(mlir::Block *originst) const {
+  if (!originalToNewFn.contains(originst)) {
+    llvm::errs() << oldFunc << "\n";
+    llvm::errs() << newFunc << "\n";
+    llvm::errs() << originst << "\n";
+    llvm_unreachable("Could not get new blk from original");
+  }
+  return originalToNewFn.lookupOrNull(originst);
+}
+
+Operation * mlir::enzyme::MGradientUtilsReverse::getNewFromOriginal(Operation *originst) const {
+  auto found = originalToNewFnOps.find(originst);
+  if (found == originalToNewFnOps.end()) {
+    llvm::errs() << oldFunc << "\n";
+    llvm::errs() << newFunc << "\n";
+    for (auto &pair : originalToNewFnOps) {
+      llvm::errs() << " map[" << pair.first << "] = " << pair.second << "\n";
+    }
+    llvm::errs() << originst << " - " << *originst << "\n";
+    llvm_unreachable("Could not get new op from original");
+  }
+  return found->second;
+}
+
+Operation *mlir::enzyme::MGradientUtilsReverse::cloneWithNewOperands(OpBuilder &B, Operation *op) {
+  BlockAndValueMapping map;
+  for (auto operand : op->getOperands())
+    map.map(operand, getNewFromOriginal(operand));
+  return B.clone(*op, map);
+}
+
+bool mlir::enzyme::MGradientUtilsReverse::isConstantValue(Value v) const {
+  if (isa<mlir::IntegerType>(v.getType()))
+    return true;
+  if (isa<mlir::IndexType>(v.getType()))
+    return true;
+
+  if (matchPattern(v, m_Constant()))
+    return true;
+
+  // TODO
+  return false;
+}
+
+Value mlir::enzyme::MGradientUtilsReverse::invertPointerM(Value v, OpBuilder &builder){
+  // TODO
+  if (invertedPointers.contains(v)){
+    assert(onlyUsedInParentBlock(v));
+    return invertedPointers.lookupOrNull(v);
+  }
+
+  if(invertedPointersGlobal.contains(v)){
+    Value gradient = invertedPointersGlobal.lookupOrNull(v);
+    Type type = gradient.getType();
+    if (GradientType gType = dyn_cast<GradientType>(type)) {
+      Value ret = builder.create<enzyme::GetGradientOp>(v.getLoc(), gType.getBasetype(), gradient);
+      return ret;
+    }
+    else{
+      llvm_unreachable("found invalid type");
+    }
+  }
+
+  if (isConstantValue(v)) {
+    llvm_unreachable("invert pointer of constant value");
+  }
+  llvm::errs() << " could not invert pointer v " << v << "\n";
+  llvm_unreachable("could not invert pointer");
```

Comment:
I see some code above checking the result of `invertPointerM` for nullity, but we never actually return null from this function. We should discuss the error handling approach: either we assume this never fails and the check in the caller is unnecessary, or we return nulls from here and handle errors more gracefully.